### PR TITLE
EIP-7251:  Add EL triggered consolidations

### PR DESCRIPTION
**Pull request by rjnrohit**
## Changes
This PR implements : https://github.com/ethereum/EIPs/pull/8625. more info at: https://eips.ethereum.org/EIPS/eip-7251 and https://github.com/ethereum/execution-apis/pull/554

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs [hunk 1]
```diff
@@ -26,6 +27,9 @@ public class ConsensusRequest
     [JsonIgnore]
     public byte[]? PubKeyField { get; set; }
 
+    [JsonIgnore]
+    public byte[]? TargetPubkeyField { get; set; }
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs:31]: to consider: you could reuse WithdrawalCredentialsField, of course with consider naming as well

**Review by MarekM25 — APPROVED** [review]

File: src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs [hunk 2]
```diff
@@ -38,46 +39,57 @@ public class ConsensusRequest
...
+                consolidationRequestCount++;
+            }
         }
 
-        return (depositCount, withdrawalRequestCount);
+        return (depositCount, withdrawalRequestCount, consolidationRequestCount);
     }
 
-    public static (Deposit[]? deposits, WithdrawalRequest[]? withdrawalRequests) SplitRequests(this ConsensusRequest[]? requests)
+    public static (Deposit[]? deposits, WithdrawalRequest[]? withdrawalRequests, ConsolidationRequest[]? consolidationRequests) SplitRequests(this ConsensusRequest[]? requests)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs:67]: out of scope of this PR: I am thinking how can we make this more general and easily extendable?

File: src/Nethermind/Nethermind.Merge.Plugin/Data/ExecutionPayloadV4.cs [hunk 3]
```diff
@@ -29,25 +29,32 @@ protected new static TExecutionPayload Create<TExecutionPayload>(Block block) wh
         {
             executionPayload.DepositRequests = Array.Empty<Deposit>();
             executionPayload.WithdrawalRequests = Array.Empty<WithdrawalRequest>();
+            executionPayload.ConsolidationRequests = Array.Empty<ConsolidationRequest>();
         }
         else
         {
-            (int depositCount, int withdrawalRequestCount) = blockRequests.GetTypeCounts();
+            (int depositCount, int withdrawalRequestCount, int consolidationRequestCount) = blockRequests.GetTypeCounts();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Data/ExecutionPayloadV4.cs:36]: this seems like same code in `SplitRequests`?

File: src/Nethermind/Nethermind.Merge.Plugin/Data/ExecutionPayloadBodyV2Result.cs [hunk 4]
```diff
@@ -14,12 +14,14 @@ public class ExecutionPayloadBodyV2Result : ExecutionPayloadBodyV1Result
         IList<Transaction> transactions,
         IList<Withdrawal>? withdrawals,
         IList<Deposit>? deposits,
-        IList<WithdrawalRequest>? withdrawalsRequests
+        IList<WithdrawalRequest>? withdrawalsRequests,
+        IList<ConsolidationRequest>? consolidationRequests
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Data/ExecutionPayloadBodyV2Result.cs:18]: migrate (all) to IReadOnlyList

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V4.cs [hunk 5]
```diff
@@ -320,25 +320,32 @@ public virtual async Task
 
         Deposit[]? deposits = null;
         WithdrawalRequest[]? withdrawalRequests = null;
+        ConsolidationRequest[]? consolidationRequests = null;
 
         if (requests is not null)
         {
-            (int depositCount, int withdrawalRequestCount) = requests.GetTypeCounts();
+            (int depositCount, int withdrawalRequestCount, int consolidationRequestCount) = requests.GetTypeCounts();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.V4.cs:327]: This looks like duplicate of SplitRequests code?

**Review by LukaszRozmej — APPROVED** [review]
This can be done in next PR, but I think we should start leveraging ConsensusRequest polimorphism, so we can easily add more in the future with lesser code

File: src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs (hunk 2 above)
Comment (rjnrohit) [reply src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs:67]: will experiment in different PR

File: src/Nethermind/Nethermind.Merge.Plugin/Data/ExecutionPayloadBodyV2Result.cs (hunk 4 above)
Comment (rjnrohit) [reply src/Nethermind/Nethermind.Merge.Plugin/Data/ExecutionPayloadBodyV2Result.cs]: Done, thanks!

File: src/Nethermind/Nethermind.Specs/Forks/00_Olympic.cs [hunk 6]
```diff
@@ -28,6 +28,7 @@ protected Olympic()
             ValidateReceipts = true;
 
             // The below addresses are added for all forks, but the given EIPs can be enabled at a specific timestamp or block.
+            Eip7251ContractAddress = Eip7251Constants.ConsolidationRequestPredeployAddress;
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Specs/Forks/00_Olympic.cs:31]: becuase you don't add them for transition. You don't need to have transition to specify address. We could seperate it to fork-agnostic properties

**Pull request merged**